### DISCUSSION TITLE
Why am I getting an error on build on one site, but not the other?  I am using the same code!

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#63782] [Help] [answered]
### Summary

I am guessing I am getting this error given the different versions of Next that I am running??

version 13.1.1 seems to be OK
version  13.5.4  is NOT OK

Error is this one:
```
./components/chat/message-input.js
55:6  Warning: React Hook useEffect has a missing dependency: 'conversationId'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps
102:6  Warning: React Hook useEffect has a missing dependency: 'emailChatHistory'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps
122:6  Warning: React Hook useEffect has a missing dependency: 'emailChatHistory'. Either include it or remove the dependency array.  react-hooks/exhaustive-deps

```

^^^ Cool, I add the dependencies... now I get this other error.

`The 'emailChatHistory' function makes the dependencies of useEffect Hook (at line 102) change on every render. To fix this, wrap the definition of 'emailChatHistory' in its own useCallback() Hook.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)`

^^^ Now how do I fix that??

The function looks like this:

  ```
const emailChatHistory = async () => {
    await fetch("/api/chat-email", {
      method: "POST",
      headers: {
        Accept: "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(chatToEmailServer),
      keepalive: true,
    }).then((res) => {
      console.log("response received");
      if (res.status === 200) {
        console.log("response succeeded");
      }
    });
  };
```

Also... I want to update the website to the newer Next version, but need to resolve these errors.

### Additional information

```js
The full code is here:
https://github.com/Elindo586/nextprojectd/blob/main/components/chat/message-input.js
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Please do not do the eslint disable, it'll cause even more issues. The lint warning is there for a reason!

The key here is understanding JavaScript pass by reference, how React renders, and how `Object.is` is used to test equality in Effects, and other places.

The short answer/recommendation, since your `emailChatHistory` function needs `chatToEmailServer`, and you do that through a closure, you can refactor to do it as a parameter:

1. Move `emailChatHistory` out of your function component, say at line 20, after the imports and before the component.
2. Rework `emailChatHistory` to:
```ts
  const emailChatHistory = (chatToEmailServer) => {
   return fetch("/api/chat-email", {
      method: "POST",
      headers: {
        Accept: "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(chatToEmailServer),
      keepalive: true,
    }).then((res) => {
      console.log("response received");
      if (res.status === 200) {
        console.log("response succeeded");
      }
    });
  };
```
3. Now update the places where `chatToEmailServer` is used.

Long answer:

The change above turns `emailChatHistory` into a so-called stable reference, which means, it'll always be the same. 

Consider this:

```js
const createFoo = () => {
   const foo = () => {}
   return foo
}

console.log(Object.is(createFoo(), createFoo())) // What do you think this logs...
```

The `foo` function is allocated, and a reference to it is created, which you hold in `foo`, when you compare it with another reference, it will be false, because it is also a different function allocation. Even though they look the same, and come from the same function factory.

Similarly, when React runs your Component, a new `emailChatHistory` function is allocated, and when React compares that in the `useEffect`, to the previous `chatToEmailServer` it will see a different function, and the effect runs again.

As you have it, and not with the recommended way above, not having the function within the effect might work, but it often causes bugs, or unexpected behavior, as the codebase changes overtime. 

Another option would've been to, wrap `chatToEmailServer` in a `useCallback`, that'll tell React to save the function reference, given the dependencies:

```js
const emailChatHistory = useCallback(() => {
 /* fetch implementation */
}, [chatToEmailServer])
``` 

It is important to understand these warnings, they are often a hint that, something can be done differently, or that the codebase is moving in the wrong direction.

**Reply by aziyatali** [reply]
I will be more careful with my responses haha. Nice response.

**Reply by Elindo586** [reply]
Ok, this has been solved using useCallback, plus using useMemo on a variable use in the useCallback.

But of course there are other problems that I am working with.

It is interesting that different versions of Next or React might work differently given the same code.  

For now I am getting a type error from an api page when I send the information, and the same code is working fine in a different REACT/Next version..

**Reply by icyJoseph** [reply]
Maybe open a different issue/question for that matter. I don't think it is necessary an issue with versions, but let us see.
